### PR TITLE
Pass measurement count to `Flp.decode()`

### LINKED ISSUE #35: Full FLP support: Pass `num_measurements` to `Flp.decode()`
Some FLPs require knowing the number of measurements in order to [decode the aggregate result](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-vdaf#section-7.1.1-6.1), e.g., when computing the arithmetic mean.

[We don't support this at the moment](https://github.com/jimouris/draft-mouris-cfrg-mastic/blob/077fb76db8a0f62f8391cb44c27c89dce05fd7ca/poc/vdaf_mastic.py#L189-L199) because we don't count how many inputs match each candidate prefix.

One way to address this to always aggregate prefix counts [the way we do for plain heavy-hitters](https://datatracker.ietf.org/doc/html/draft-mouris-cfrg-mastic#section-5.3-7).

### PR DESCRIPTION
**Pull request by cjpatton**
Closes #35.
Stacked on #43 (merge that first).

At the moment we simply pass `meas_count == 0` to `Flp.decode()` because we don't know the actual value of `meas_count`. This makes `Mastic` incompatible with any FLP that requires the measurement count at decoding time.

To address this, plumb a counter for each measurement from VIDPF evaluation to FLP unsharding. The value of the counter is `1` if the prefix is on the `alpha`-path and `0` otherwise. To accomplish this, we do the simplest thing: prepend a field element to each correction word.

Along the way, fix a bug in `Mastic.aggregate()` and `Mastic.unshard()` whereby the accumulator vector was initialized to the wrong length.

### CODE REVIEW

File: poc/vidpf.py [hunk 1]
```diff
@@ -154,8 +154,19 @@ def eval(cls,
...
 
+        # Compute the Aggregator's share of `beta`.
+        y0 = prefix_tree_share[(0, 0)][2]
+        y1 = prefix_tree_share[(1, 0)][2]
+        beta_share = vec_add(y0, y1)[1:]  # first element is the counter
+        if agg_id == 1:
+            beta_share = vec_neg(beta_share)
+
+        # Compute the Aggregator's share of the counter.
+        counter_share = y0[0] + y1[0] + cls.Field(agg_id)
```
Comment (cjpatton) [inline poc/vidpf.py:165]: @jimouris please check carefully that this matches PLASMA.

Comment (jimouris) [reply poc/vidpf.py:165]: You are referring to this specific line, right? `counter_share = y0[0] + y1[0] + cls.Field(agg_id)`

Which PLASMA check do you have in mind?

Comment (cjpatton) [reply poc/vidpf.py:165]: > You are referring to this specific line, right? `counter_share = y0[0] + y1[0] + cls.Field(agg_id)`

Yes.

> Which PLASMA check do you have in mind?

This is the thing where we make sure the value programmed into the VIDPF (i.e., part of $\beta$) is `1`.

Comment (jimouris) [reply poc/vidpf.py:165]: Not exactly. In PLASMA we do the following checks:
* for the root level: we check that $H(1 - y_{S0}^0 - y_{S0}^1) == H(y_{S1}^0 + y_{S1}^1)$
* for all other levels: we check that $H(y_{S0}^p - y_{S0}^{p||0} - y_{S0}^{p||1}) == H(-y_{S1}^p + y_{S1}^{p||0} + y_{S1}^{p||1})$

So the check that actually checks if beta is one is the first one for the root level.

**Comment by cjpatton** [comment]
Rebased.

File: poc/vidpf.py (hunk 1 above)
Comment (cjpatton) [reply poc/vidpf.py:165]: I believe that's exactly what I'm doing here:
```python
counter_share = y0[0] + y1[0] + cls.Field(agg_id)
```
Except that the second server's share is not negated. (For some reason, the math works out this way in the current VIDPF implementation, as we noted on an earlier PR.)

**Review by jimouris — APPROVED** [review]

File: poc/vidpf.py (hunk 1 above)
Comment (hannahdaviscrypto) [inline poc/vidpf.py:165]: I think this counter_share isn't getting returned by the VIDPF or included in the out_share?

**Review by hannahdaviscrypto — APPROVED** [review]
One possible bug in how the counter_share is used, but otherwise looks good to me. Thanks, Chris!

File: poc/vidpf.py (hunk 1 above)
Comment (cjpatton) [reply poc/vidpf.py:165]: the `out_share` returned at the end of this function would include a share of  $1 \|| \beta$ if one of the prefixes is on the $\alpha$ path. At this point we're just checking that the programmed value is indeed $1$.

**Pull request merged**